src/dmd/statementsem.d

```diff
@@ -728,12 +728,16 @@ package (dmd) extern (C++) final class StatementSemanticVisitor : Visitor
         Dsymbol sapply = null;                  // the inferred opApply() or front() function
         if (!inferForeachAggregate(sc, fs.op == TOK.foreach_, fs.aggr, sapply))
         {
-            const(char)* msg = "";
-            if (fs.aggr.type && isAggregate(fs.aggr.type))
+            if (!oaggr.type)
             {
-                msg = ", define `opApply()`, range primitives, or use `.tupleof`";
+                fs.error("invalid `foreach` aggregate `%s`", oaggr.toChars());
```

Comment:
When is this branch taken?

Reply:
Not sure. I originally kept it because the previous code checked whether the type was set. Replaced it with an `assert` now.